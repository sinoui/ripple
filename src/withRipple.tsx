import React from 'react';
import { findDOMNode } from 'react-dom';
import { RippleConfig } from './types';
import RippleElement from './RippleElement';

interface BaseProps {
  disabled?: boolean;
}

const withRipple = (rippleConfig?: RippleConfig) => {
  function createEnableRippleComponent<P extends BaseProps>(Component: any) {
    const name =
      typeof Component === 'string' ? Component : Component.displayName;
    return class EnableRippleComponent extends React.Component<P> {
      // eslint-disable-next-line react/static-property-placement
      public static displayName = `withRipple(${name})`;

      public static sinouiName = name;

      public rippleElement?: RippleElement;

      public componentDidMount() {
        this.createRippleElement();
      }

      public componentDidUpdate(prevProps: P) {
        const { disabled } = this.props;
        const disabledChanged = prevProps.disabled !== disabled;

        if (disabledChanged && disabled && this.rippleElement) {
          this.rippleElement.disable();
        } else if (disabledChanged && !disabled) {
          if (!this.rippleElement) {
            this.createRippleElement();
          } else {
            this.rippleElement.enable();
          }
        }
      }

      public componentWillUnmount() {
        if (this.rippleElement) {
          this.rippleElement.cancel();
        }
      }

      public createRippleElement() {
        // eslint-disable-next-line react/no-find-dom-node
        const element = findDOMNode(this) as HTMLElement;
        this.rippleElement = new RippleElement(element, rippleConfig);
        requestAnimationFrame(() => {
          const { position } = window.getComputedStyle(element);
          if (!position || position === 'static') {
            element.style.position = 'relative';
          }
        });
      }

      public render() {
        return <Component {...this.props} />;
      }
    };
  }

  return createEnableRippleComponent;
};

export default withRipple;
