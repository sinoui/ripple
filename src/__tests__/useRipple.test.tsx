import React from 'react';
import renderer from 'react-test-renderer';
import withRipple from '../withRipple';

const Demo = () => <div>test</div>;
const RippleDemo = withRipple({})(Demo);

it('快照测试', () => {
  const tree = renderer.create(<RippleDemo />);

  expect(tree.toJSON()).toMatchSnapshot();
});
