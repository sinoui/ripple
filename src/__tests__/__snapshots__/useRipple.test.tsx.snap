// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`快照测试 1`] = `
<div>
  test
</div>
`;
